### PR TITLE
Feature/path3

### PR DESCRIPTION
**Pull request by rancescolon**
added the file handler to views.py

### CODE REVIEW

**Review by sumanyuj — DISMISSED** [review]
LGTM!

File: evergreen/eg_app/views.py [hunk 1]
```diff
@@ -12,6 +18,54 @@
...
+
+    allowedType = {'.css', '.html', '.js', '.png', '.jpg', '.jpeg', '.gif', '.mp3', '.mp4', '.xml', '.json', '.pdf','.ico'} # can add more, 
+
+    if not path.suffix.lower() in allowedType: # to deal with user uploads 
+        return HttpResponseNotFound("404 - File type not allowed")
+    
+    # #deal with /../ attacks 
+    # rootPath = Path(settings.STATIC_ROOT).resolve() #.resolve get absolute path 
+    # if not path.resolve().is_relative_to(rootPath):
+    #     return HttpResponseNotFound("404 - Invalid file path")
```
Comment (WeirdAlex03) [inline evergreen/eg_app/views.py]: Why is this added commented out? Same for sanitizedFilePath above

File: evergreen/eg_app/views.py [hunk 2]
```diff
@@ -12,6 +18,54 @@
 def index(request):
     return render(request,'index.html')
 
+def getFileType(filePath):
+    return mimetypes.guess_type(filePath)[0]
```
Comment (WeirdAlex03) [inline evergreen/eg_app/views.py]: I don't think ignoring the encoding will work for UTF-8 (i.e. emoji support). That should be appended if given

**Review by WeirdAlex03 — CHANGES_REQUESTED** [review]

File: evergreen/eg_app/views.py (hunk 2 above)
Comment (Eric-Butcher) [reply evergreen/eg_app/views.py]: the index into [0] for first index is not immediately obvious. Please assign that list index to a named variable so there is not a magic value.

File: evergreen/eg_app/views.py [hunk 3]
```diff
@@ -12,6 +18,54 @@
...
+def getFileType(filePath):
+    return mimetypes.guess_type(filePath)[0]  
+
+def fileHandler(request, fileName): # can handle img and text 
+    # sanitizedfileName = quote(fileName) # get a clean full file path 
+    # path = Path(settings.STATIC_ROOT) / sanitizedfileName
+
+    path = fileName
+
+    allowedType = {'.css', '.html', '.js', '.png', '.jpg', '.jpeg', '.gif', '.mp3', '.mp4', '.xml', '.json', '.pdf','.ico'} # can add more,
```
Comment (Eric-Butcher) [inline evergreen/eg_app/views.py]: add type hint for set[str]

File: evergreen/eg_app/views.py [hunk 4]
```diff
@@ -12,6 +18,54 @@
...
     return render(request,'index.html')
 
+def getFileType(filePath):
+    return mimetypes.guess_type(filePath)[0]  
+
+def fileHandler(request, fileName): # can handle img and text 
+    # sanitizedfileName = quote(fileName) # get a clean full file path 
+    # path = Path(settings.STATIC_ROOT) / sanitizedfileName
+
+    path = fileName
```
Comment (Eric-Butcher) [inline evergreen/eg_app/views.py]: add type hint. Is this a pure string filename or a pathlib path?

File: evergreen/eg_app/views.py [hunk 5]
```diff
@@ -12,6 +18,54 @@
...
+        
+        try:
+            with open(path, 'rb') as file:
+                content = file.read()
+
+            response = HttpResponse(content, content_type=contentType)
+            response['X-Content-Type-Options'] = "nosniff"
+            #response['Content-Length'] = str(len(content)) is need for very large file, django handles is not
+
+            print(response)
```
Comment (Eric-Butcher) [inline evergreen/eg_app/views.py]: Why are we printing this to the terminal? Instead of doing printing for debugging (assuming this is what you are doing) it is probably better to use the logging module so we can turn these on and off.

**Review by Eric-Butcher — DISMISSED** [review]

File: evergreen/eg_app/views.py (hunk 1 above)
Comment (rancescolon) [reply evergreen/eg_app/views.py]: sanitizedFilePath gets the absolute path
these lines deal with the user trys to access a different directory with a /../ attack

File: evergreen/eg_app/views.py [hunk 6]
```diff
@@ -12,6 +20,61 @@
...
+            return response
+
+        except OSError:  # catch most of them, like FileNotFoundError
+            return HttpResponseNotFound("404 Not Found")
+            # return render(request, '404.html', status=404) when making 404 pages
+    else:
+        return HttpResponseNotFound("404 Not Found")
+
+
+def addCookies(response, cookies): # add all cookies from the give dic to the response
```
Comment (WeirdAlex03) [inline evergreen/eg_app/views.py]: One more set of type hints (and make it a docstring comment)
```suggestion
def addCookies(response: HttpResponse, cookies: dict[str, str]): 
    """ Add all cookies from the given dict to the response """
```

**Review by WeirdAlex03 — CHANGES_REQUESTED** [review]

File: evergreen/eg_app/views.py [hunk 7]
```diff
@@ -12,6 +20,61 @@
 def index(request):
     return render(request,'index.html')
 
+def getFileType(filePath) -> tuple[str|None, str|None]:
```
Comment (Eric-Butcher) [inline evergreen/eg_app/views.py]: I am assuming filePath is a str? Or is it a pathlib.Path? Type annotation here would be nice

File: evergreen/eg_app/views.py [hunk 8]
```diff
@@ -12,6 +20,61 @@
...
+    contentType = mimetypes.guess_type(filePath) #returns tuple {type, encoding}
```
Comment (Eric-Butcher) [inline evergreen/eg_app/views.py:22]: Comment is nice by type annotation would be cool too.

File: evergreen/eg_app/views.py [hunk 9]
```diff
@@ -12,6 +20,61 @@
...
+    return contentType
+
+def fileHandler(request: HttpRequest, fileName: str) -> HttpResponse:  # can handle img and text
+    # get the absolute path
+    sanitizedFileName = quote(fileName)
+    path = Path(settings.STATIC_ROOT) / sanitizedFileName
+
+    allowedType: Set[str] = {'.css', '.html', '.js', '.png', '.jpg', '.jpeg', '.gif', '.mp3', '.mp4', '.xml', '.json', '.pdf', '.ico'}  # can add more
```
Comment (Eric-Butcher) [inline evergreen/eg_app/views.py]: The python type is _set_ with a lowercase, not _Set_.

File: evergreen/eg_app/views.py [hunk 10]
```diff
@@ -12,6 +20,61 @@
...
+        return HttpResponseNotFound("404 Not Found")
+
+    if path.exists() and path.is_file(): # make sure it's not a directory
+        contentType, encoding = getFileType(str(path))
+
+        try:
+            with open(path, 'rb') as file:
+                content = file.read()
+
+            if contentType is not None:
```
Comment (Eric-Butcher) [inline evergreen/eg_app/views.py]: In Python you can simply use `if contentType:` here

File: evergreen/eg_app/views.py [hunk 11]
```diff
@@ -12,6 +20,61 @@
...
+                response = HttpResponse(content, content_type=contentType)
+            else:
+                response = HttpResponse(content)
+
+            if encoding is not None:
```
Comment (Eric-Butcher) [inline evergreen/eg_app/views.py]: In Python you can simply use `if encoding:` here

File: evergreen/eg_app/views.py [hunk 12]
```diff
@@ -12,6 +20,61 @@
...
+
+    # deal with /../ attacks
+    rootPath = Path(settings.STATIC_ROOT).resolve()
+    if not path.resolve().is_relative_to(rootPath):
+        return HttpResponseNotFound("404 Not Found")
+
+    if path.exists() and path.is_file(): # make sure it's not a directory
+        contentType, encoding = getFileType(str(path))
+
+        try:
```
Comment (Eric-Butcher) [inline evergreen/eg_app/views.py:43]: It seems you only need the try/except around the file io, bring the try except closer and leave the program logic outside the try/except block to make this more readable.

File: evergreen/eg_app/views.py [hunk 13]
```diff
@@ -1,4 +1,12 @@
+from typing import Set
```
Comment (Eric-Butcher) [inline evergreen/eg_app/views.py]: remove this

**Review by WeirdAlex03 — APPROVED** [review]

**Review by Eric-Butcher — APPROVED** [review]
